infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LogReplicationServer.java

```diff
@@ -187,42 +191,78 @@ private void handleMetadataRequest(@Nonnull RequestMsg request,
         log.info("Log Replication Metadata Request received by Server.");
 
         if (isLeader.get()) {
-            String sourceClusterId = getUUID(request.getHeader().getClusterId()).toString();
-            ReplicationSession sourceSession = ReplicationSession
-                .getDefaultReplicationSessionForCluster(sourceClusterId);
 
-            LogReplicationSinkManager sinkManager = sourceSessionToSinkManagerMap.get(sourceSession);
+            LogReplicationSession session = getSession(request);
+
+            LogReplicationSinkManager sinkManager = sessionToSinkManagerMap.get(session);
 
-            // If no sink Manager is found, drop the message and log an error
             if (sinkManager == null) {
-                log.error("Sink Manager not found for remote cluster {}.  This could be due to a topology mismatch.",
-                    getUUID(request.getHeader().getClusterId()).toString());
+                log.error("Sink Manager not found for session {}, total={}, sessions={}", session,
+                        sessionToSinkManagerMap.size(), sessionToSinkManagerMap.keySet());
                 return;
             }
 
-            LogReplicationMetadataManager metadataMgr = sinkManager.getLogReplicationMetadataManager();
+            ReplicationMetadata metadata = sessionManager.getMetadataManager().getReplicationMetadata(session, false);
+            ResponseMsg response = getMetadataResponse(request, metadata);
```

Comment:
We should add session information in the response msgs.
Here as well as for handleLrEntryRequest

Reply:
No, if request has it, then response should have it.
I think I need to complete my review to see it being populated.

Reply:
@pankti-m just for my understanding. IIUC, are there some requests without session populated in the request header?

Reply:
yes, session information does not get populated in the requests right now.